go/libraries/doltcore/sqle/dsess/session.go

```diff
@@ -223,149 +268,89 @@ func (d *DoltSession) ValidateSession(ctx *sql.Context, dbName string) error {
-
-	// TODO: this needs to happen for every DB in the database, not just the one named in the transaction
-	if sessionState != nil && sessionState.db != nil {
-		rrd, ok := sessionState.db.(RemoteReadReplicaDatabase)
-		if ok && rrd.ValidReplicaState(ctx) {
-			err := rrd.PullFromRemote(ctx)
-			if err != nil && !IgnoreReplicationErrors() {
-				return nil, fmt.Errorf("replication error: %w", err)
-			} else if err != nil {
-				WarnReplicationError(ctx, err)
-			}
-		}
-	}
-
-	if sessionState.readOnly {
-		return DisabledTransaction{}, nil
-	}
-
-	nomsRoots := make(map[string]hash.Hash)
-	for _, db := range d.provider.DoltDatabases() {
+	// Take a snapshot of the current noms root for every database under management
+	doltDatabases := d.provider.DoltDatabases()
+	txDbs := make([]SqlDatabase, 0, len(doltDatabases))
+	for _, db := range doltDatabases {
 		// TODO: this nil check is only necessary to support UserSpaceDatabase and clusterDatabase, come up with a better set of
 		//  interfaces to capture these capabilities
 		ddb := db.DbData().Ddb
 		if ddb != nil {
-			nomsRoot, err := ddb.NomsRoot(ctx)
-			if err != nil {
-				return nil, err
+			rrd, ok := db.(RemoteReadReplicaDatabase)
+			if ok && rrd.ValidReplicaState(ctx) {
+				err := rrd.PullFromRemote(ctx)
+				if err != nil && !IgnoreReplicationErrors() {
+					return nil, fmt.Errorf("replication error: %w", err)
+				} else if err != nil {
+					WarnReplicationError(ctx, err)
+				}
 			}
-			nomsRoots[strings.ToLower(db.Name())] = nomsRoot
-		}
-	}
 
-	if _, v, ok := sql.SystemVariables.GetGlobal(ReadReplicaRemote); ok && v != "" {
-		err = sessionState.dbData.Ddb.Rebase(ctx)
-		if err != nil && !IgnoreReplicationErrors() {
-			return nil, err
-		} else if err != nil {
-			WarnReplicationError(ctx, err)
+			if _, v, ok := sql.SystemVariables.GetGlobal(ReadReplicaRemote); ok && v != "" {
+				err := ddb.Rebase(ctx)
+				if err != nil && !IgnoreReplicationErrors() {
+					return nil, err
+				} else if err != nil {
+					WarnReplicationError(ctx, err)
+				}
+			}
+
+			txDbs = append(txDbs, db)
 		}
 	}
 
-	wsRef := sessionState.WorkingSet.Ref()
-	ws, err := sessionState.dbData.Ddb.ResolveWorkingSet(ctx, wsRef)
-	// TODO: every HEAD needs a working set created when it is. We can get rid of this in a 1.0 release when this is fixed
-	if err == doltdb.ErrWorkingSetNotFound {
-		ws, err = d.newWorkingSetForHead(ctx, wsRef, dbName)
-		if err != nil {
-			return nil, err
-		}
-	} else if err != nil {
+	tx, err := NewDoltTransaction(ctx, txDbs, tCharacteristic)
+	if err != nil {
 		return nil, err
 	}
 
-	// logrus.Tracef("starting transaction with working root %s", ws.WorkingRoot().DebugString(ctx, true))
+	// The engine sets the transaction after this call as well, but since we begin accessing data below, we need to set
+	// this now to avoid seeding the session state with stale data in some cases. The duplication is harmless since the
+	// code below cannot error, but it would to get rid of
```

Comment:
(minor) looks like the end of this comment go cut off? Should this be  "it would be hard to get rid of"?

Reply:
I think it was supposed to be "but it would be nice to get rid of" but I rephrased to make it clearer